### ISSUE TITLE
Transformer Lack of Embedding Layer and Positional Encodings

### ISSUE THREAD

**Issue by TrentBrick** [pytorch/pytorch#24826]
The Transformer implementation docs (https://pytorch.org/docs/stable/nn.html?highlight=transformer#torch.nn.Transformer) state that they implement the original paper but fail to acknowledge that they don’t implement the following: 
* Layer Norm as the default normalization option. 
* Positional Encodings
* Embeddings before the encoding and decoding step 

It’s fine that these are all not implemented directly in the module but making it more clear that they aren’t and were in the original paper would be helpful.

**Comment by jeffreyksmithjr** [comment]
Tagging this to reflect that we could implement the functionality described or just document the existing implementation. Both would seem to be somewhat valid.

**Comment by andreaskoepf** [comment]
>     * Layer Norm as the default normalization option.

Could you elaborate what you are missing regarding `LayerNorm` as default normalization? 
[TransformerEncoderLayer](
https://github.com/pytorch/pytorch/blob/c6fe864db3e17830bf12957a64e6fd579ddeffad/torch/nn/modules/transformer.py#L223) and [TransformerDecoderLayer](https://github.com/pytorch/pytorch/blob/c6fe864db3e17830bf12957a64e6fd579ddeffad/torch/nn/modules/transformer.py#L302) both use `LayerNorm` (hard built-in) and the `nn.Transformer` creates a `TransformerEncoder` with additional final `LayerNorm` by default.

I think the Transformer abstraction that was implemented is really good as it is. Maybe [fairseq](https://github.com/pytorch/fairseq) could be mentioned for people looking for full NLP models which offer embedding to decoding.

**Comment by TrentBrick** [comment]
Re LayerNorm: Huh, sorry that I missed that. I guess I got confused by the fact that there is an option for further normalization in the "TransformerEncoder" and Decoder. 

![image](https://user-images.githubusercontent.com/12433427/63453782-6dfb7a00-c417-11e9-927d-514777fdab61.png)

I don't think anything more needs to be implemented just better documentation on what people are and aren't getting.

Embeddings aren't a big deal but were used in the original paper and I think it should be flagged that they aren't here. 

The same is true for the positional encodings but these are much more important for the model's performance and difficult to implement. (this was noted here https://github.com/pytorch/pytorch/issues/10459#issuecomment-413116713)

**Comment by Vichoko** [comment]
Recently [found this snippet in here](https://towardsdatascience.com/how-to-code-the-transformer-in-pytorch-24db27c8f9ec#d554) that implements PositionalEncoding that can be easily added at the beggining of your `forward(x)` and before calling the transformer encoder forward.
```

class PositionalEncoder(nn.Module):
    def __init__(self, d_model, max_seq_len = 80):
        super().__init__()
        self.d_model = d_model
        
        # create constant 'pe' matrix with values dependant on 
        # pos and i
        pe = torch.zeros(max_seq_len, d_model)
        for pos in range(max_seq_len):
            for i in range(0, d_model, 2):
                pe[pos, i] = \
                math.sin(pos / (10000 ** ((2 * i)/d_model)))
                pe[pos, i + 1] = \
                math.cos(pos / (10000 ** ((2 * (i + 1))/d_model)))
                
        pe = pe.unsqueeze(0)
        self.register_buffer('pe', pe)
 
    
    def forward(self, x):
        # make embeddings relatively larger
        x = x * math.sqrt(self.d_model)
        #add constant to embedding
        seq_len = x.size(1)
        x = x + Variable(self.pe[:,:seq_len], \
        requires_grad=False).cuda()
        return x
```

**Comment by volpato30** [comment]
> Recently [found this snippet in here](https://towardsdatascience.com/how-to-code-the-transformer-in-pytorch-24db27c8f9ec#d554) that implements PositionalEncoding that can be easily added at the beggining of your `forward(x)` and before calling the transformer encoder forward.
> 
> I updated to work for more recent versions:
> 
> ```python
> class PositionalEncoder(torch.nn.Module):
>     def __init__(self, d_model, max_seq_len=160):
>         super().__init__()
>         self.d_model = d_model
>         pe = torch.zeros(max_seq_len, d_model)
>         for pos in range(max_seq_len):
>             for i in range(0, d_model, 2):
>                 pe[pos, i] = \
>                     math.sin(pos / (10000 ** ((2 * i) / d_model)))
>                 pe[pos, i + 1] = \
>                     math.cos(pos / (10000 ** ((2 * (i + 1)) / d_model)))
>         pe = pe.unsqueeze(0)
>         self.register_buffer('pe', pe)
> 
>     def forward(self, x):
>         with torch.no_grad():
>             x = x * math.sqrt(self.d_model)
>             seq_len = x.size(1)
>             pe = self.pe[:, :seq_len]
>             x = x + pe
>             return x
> ```
> 
> Edit: Surely that _nested for_ can be optimized with tensor logic because for big sequences or big `d_model ` parameter it can take some time to initialize the module.

May I ask why use no_grad in the forward function? Wouldn't it prevent the update of word embeddings?

**Comment by kryptec** [comment]
This tutorial uses the TransformerEncoder class to implement the original paper, and looks like it has a cleaner positional encoding. 

https://pytorch.org/tutorials/beginner/transformer_tutorial.html

**Comment by mmwebster** [comment]
+1 on it being odd that nn.Transformer doesn't implement positional encoding. The lack of word embeddings makes sense, but it seems like the vast majority of transformers would require positional encodings.

Also, example usage for nn.Transformer training and evaluation would improve the docs a lot. It seems like nn.Transformer has been abandoned in favor of the encoder, decoder, etc. classes that comprise it, as suggested by the tutorial that @kryptec referenced. Seems like it would be better to have a tutorial for nn.Transformer, and details on how to override pieces of functionality, like how positional encoding is performed.

**Comment by Vichoko** [comment]
Yeah, right. Also it could be useful to provide an implementation of new alternative positional encodings, for example, relative positional encodings.

**Comment by david-waterworth** [comment]
I agree positional encoding should really be implemented and part of the transformer - I'm less concerned that the embedding is separate.

In particular, the input shape of the PyTorch transformer is different from other implementations (src is SNE rather than NSE) meaning you have to be very careful using common positional encoding implementations.

I also find it a little odd/confusing that src is SNE yet src_key_padding_mask is NS

**Comment by jbschlosser** [comment]
Bumping to high pri due to activity & because positional encodings are generally essential for using the Transformer layers effectively

**Comment by narain1** [comment]
can we just push the positional encoding from [pytorch transformer tutorial](https://pytorch.org/tutorials/beginner/transformer_tutorial.html) to the nn module?

**Comment by granitdula** [comment]
> can we just push the positional encoding from [pytorch transformer tutorial](https://pytorch.org/tutorials/beginner/transformer_tutorial.html) to the nn module? and I find it to be in terms with the paper

The tutorial here seems to create pe with the shape (max_len, 1, d_model), and later, when its added to the input x which has shape (S, N, E), its sliced so its first dimension is S as well, giving it the shape (S, 1, d_model). But this of course doesn't work if you are using any batch size, N, that is not 1. Wouldn't this `PositionalEncoding` need the batch size as an argument so you can then update pe, at the end of the `__init__`, to:

`pe = pe.repeat(1, batch_size, 1)`

But other than this, I'm also in agreement with having positional encoding included by default in the nn.Transformer class.

**Comment by aohan237** [comment]
vote for this feature

**Comment by ankit61** [comment]
I too agree that positional encodings should be a part of the transformer class. However, instead of hard coding them, it is perhaps better to pass a flag in the constructor about whether to use positional encodings or not (can default to true). There definitely can be (and are) cases when positional encodings are not necessary, so such flexibility would be important

**Comment by david-waterworth** [comment]
I agree with @ankit61 - many transformer models use position embeddings in place of the original sinusoidal encoding, and some may use no encoding. So the actual embedding layer should be configurable

**Comment by nkalpakis21** [comment]
i see this has been open for awhile. is there trouble getting a dev to fix this issue? what is blocking this for so long

**Comment by cajmorgan** [comment]
I have a similar question here, this thread is now about 7 years old and I still can't see any updates to this
